go/test/server_test.go

```diff
@@ -0,0 +1,240 @@
+		t.Fatal(err)
+	}
+	defer c.Close(websocket.StatusInternalError, "the sky is falling")
+
+	StartGrpc(c, 1, "main.Greeter/UnaryHello", make(map[string]string), ctx)
+	data, err := proto.Marshal(&HelloRequest{
+		Name: "boris 2",
+	})
+	//todo make this a class for easier tests
+	MessageGrpc(c, 1, data, ctx)
+	CompleteGrpc(c, 1, ctx)
+	log.Info("Reading response")
+	_, r, err := c.Reader(ctx)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	response2 := &GrpcFrame{}
+	bytesValue, err := ioutil.ReadAll(r)
+	if err := proto.Unmarshal(bytesValue, response2); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+
+	serverResponse := &HelloReply{}
+	if err := proto.Unmarshal(response2.GetBody().GetData(), serverResponse); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+	log.Infof("got server response %v", response2)
+	c.Close(websocket.StatusNormalClosure, "")
+}
+
+func Test_streamCall(t *testing.T) {
+	log := logrus.New()
+	t.Parallel()
+	go startGrpcServer()
+	server, err := NewEchoServer(logrus.NewEntry(log), "localhost:50051")
+	s := httptest.NewServer(server)
+
+	opt := []grpc.DialOption{}
+	opt = append(opt, grpc.WithCodec(proxy.Codec()))
+	opt = append(opt, grpc.WithInsecure())
+
+	log.Info("started server")
+	defer s.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	c, _, err := websocket.Dial(ctx, s.URL, &websocket.DialOptions{
+		Subprotocols: []string{"grpc-websocket"},
+	})
+	log.Info("got websocket")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer c.Close(websocket.StatusInternalError, "the sky is falling")
+
+	StartGrpc(c, 1, "main.Greeter/BiStreamHello", make(map[string]string), ctx)
+	data, err := proto.Marshal(&HelloRequest{
+		Name: "boris 2",
+	})
+	//todo make this a class for easier tests
+	MessageGrpc(c, 1, data, ctx)
+	CompleteGrpc(c, 1, ctx)
+	log.Info("Reading response")
+	_, r, err := c.Reader(ctx)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	response2 := &GrpcFrame{}
+	bytesValue, err := ioutil.ReadAll(r)
+	if err := proto.Unmarshal(bytesValue[5:], response2); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+
+	serverResponse := &HelloReply{}
+	if err := proto.Unmarshal(response2.GetBody().GetData(), serverResponse); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+
+	log.Infof("got server response %v", serverResponse)
+
+	log.Info("Reading response2")
+	_, r, err = c.Reader(ctx)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	response2 = &GrpcFrame{}
+	bytesValue, err = ioutil.ReadAll(r)
+
+	log.Infof("response size %v", len(bytesValue))
+	if err := proto.Unmarshal(bytesValue[5:], response2); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+
+	serverResponse = &HelloReply{}
+	if err := proto.Unmarshal(response2.GetBody().GetData(), serverResponse); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+	log.Infof("got server response2 %v", serverResponse)
+	c.Close(websocket.StatusNormalClosure, "")
+}
+
+func CompleteGrpc(ws *websocket.Conn, streamId uint32, ctx context.Context) error {
```

Comment:
How is the stream id managed? I might of missed that.

Reply:
The client will give streamid's back and has to manage it properly.
I would use an incremental id and assume that you don't do 4g requests. 
A max number of streams per channel might be useful as protection.